airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/abstract_stream.py

```diff
@@ -0,0 +1,103 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, Optional, Tuple, Union
+
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+from deprecated.classic import deprecated
+
+PrimaryKey = Union[str, List[str]]
+
+
+@deprecated("This class is experimental. Use at your own risk.")
+class AbstractStream(ABC):
+    """
+    AbstractStream is an experimental interface for streams developed as part of the Concurrent CDK.
+    This interface is not yet stable and may change in the future. Use at your own risk.
+
+    Why create a new interface instead of adding concurrency capabilities the existing Stream?
+    We learnt a lot since the initial design of the Stream interface, and we wanted to take the opportunity to improve.
+
+    High level, the changes we are targeting are:
+    - Removing superfluous or leaky parameters from the methods' interfaces
+    - Using composition instead of inheritance to add new capabilities
+
+    To allow us to iterate fast while ensuring backwards compatibility, we are creating a new interface with a facade object that will bridge the old and the new interfaces.
+    Source connectors that which to leverage concurrency need to implement this new interface. An example will be available shortly
+
+    Current restrictions on sources that implement this interface. Not all of these restrictions will be lifted in the future, but most will as we iterate on the design.
+    - Only full refresh is supported. This will be addressed in the future.
+    - The read method does not accept a cursor_field. Streams must be internally aware of the cursor field to use. User-defined cursor fields can be implemented by modifying the connector's main method to instantiate the streams with the configured cursor field.
+    - Streams cannot return user-friendly messages by overriding Stream.get_error_display_message. This will be addressed in the future.
+    - The Stream's behavior cannot depend on a namespace
+    - TypeTransformer is not supported. This will be addressed in the future.
+    - Nested cursor and primary keys are not supported
+    """
+
+    @abstractmethod
+    def read(self) -> Iterable[Record]:
```

Comment:
As I understand it this would remove our ability to interleave records and log messages and have them emitted in a specific order - is that right?

If so - is there a strong reason to impose this restriction upon ourselves?

Reply:
We can still emit log messages using a message repository. [This follow up PR has an example of this in ThreadBasedConcurrentStream's implementation of the read method](https://github.com/airbytehq/airbyte/pull/30113/files#diff-4318f9190b9186c7d0fab671dc3f4441f9d22a87cb970ec6fa1578995bff69beR107)

Reply:
Yes but the order won't be preserved, right? I know that with the concurrent CDK the order of records will be wonky anyway, just wondering if there are scenarios where we'd want to be able to preserve order of records & logs being emitted within a partition.

Reply:
I would expect the order to still be preserved since the abstract source reads from the message repository after each record https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L207

Reply:
> we can always add an option to send messages that should be processed before the record

I'm thinking more of messages that need to be sent within the read, e.g. if we were to do something in the future like emit a new catalog when we see that records don't conform to the schema, or if at some point we wanted to emit log messages in the middle of a sync (e.g. for tracking the number of items processed, or because we've reached some kind of limit, or encountered an error that we want to log but don't want to break the sync). Is the simplicity of having `read` just emit one type of message the primary benefit of changing the interface?

Reply:
> Is the simplicity of having read just emit one type of message the primary benefit of changing the interface?

The read method is intentionally not returning a message, only record objects, which will be turned into `AirbyteMessage`s by the `AbstractSource`. The `Stream.read_records` interface is odd because it returns either AirbyteMessage or a dict.

An alternative I wouldn't mind too much would be for `read` to always return `AirbyteMessage`s. That would be more consistent and would make it clear to the consumer they need to check the type of the returned objects.

There are two reasons I prefer returning an iterable of records instead of messages from the stream:
1. Most consumers only care about records. We got hit by this interface change a few time in the past, and had to add a bunch of typechecks 
- https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py#L137-L144
- https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py#L328-L335
- I think `stream_helper.get_first_record_for_slice` is still wrong because [it might return a log message](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/utils/stream_helper.py#L39-L40)...
2. I think we should emit these messages using a MessageRepository since they can be emitted from outside the read flow ([the single use oauth sends control messages for example](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py#L232))

----
tl;dr: I'm open to changing the interface to always return AirbyteMessages, but I'm worried it'll be more difficult to use

Reply:
@clnoll @maxi297 @brianjlai and I discussed this today. 

The consensus is that the stream's read method will only output records. They will use the message repository to emit other types of messages.

Further improvements also came out of the discussion to standardize how these messages are emitted. This can be tracked in [this issue](https://github.com/airbytehq/airbyte/issues/31308), but is imo lower priority.